File: sdks/python/src/devhub_sdk/_http_transport.py
from __future__ import annotations

import json
from abc import ABC, abstractmethod
from typing import Any
from urllib.error import HTTPError
from urllib.request import Request, urlopen
from uuid import uuid4

from ._json import load_json_text
from ._jsonrpc import validate_response_envelope
from .models import DevHubClientOptions, RuntimeConnectionInfo


class JsonRpcHttpTransport(ABC):
    """HTTP JSON-RPC 传输抽象。"""

    @abstractmethod
    def send(self, method: str, params: dict[str, Any] | None) -> dict[str, Any]:
        """发送 JSON-RPC 请求并返回结果载荷。"""


class UrllibJsonRpcHttpTransport(JsonRpcHttpTransport):
    """基于 urllib 的默认 HTTP JSON-RPC 传输。"""

    def __init__(self, connection_info: RuntimeConnectionInfo, options: DevHubClientOptions) -> None:
        self._connection_info = connection_info
        self._options = options.clone()

    def send(self, method: str, params: dict[str, Any] | None) -> dict[str, Any]:
        request_id = f"req-{uuid4().hex}"
        payload = {
            "jsonrpc": "2.0",
            "id": request_id,
            "method": method,
        }
        if params is not None:
            payload["params"] = params

        request = Request(
            self._connection_info.rpc_endpoint,
            data=json.dumps(payload, allow_nan=False).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {self._connection_info.token}",
                "Content-Type": "application/json",
                "X-DevHub-Protocol": str(self._options.protocol_version),
                "X-DevHub-ClientId": self._options.client_id,
                "X-DevHub-ClientSessionId": self._options.client_session_id,
            },
            method="POST",
        )

        try:
            with urlopen(request, timeout=self._options.request_timeout) as response:
                body = response.read().decode("utf-8")
        except HTTPError as exc:
            body = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"HTTP 请求失败：{exc.code} {exc.reason}，响应体：{body}") from exc

        root = load_json_text(body, source="HTTP 响应体")
        return validate_response_envelope(root, request_id)

File: sdks/javascript/src/http-transport.ts
import {
  DevHubConnectionError,
  DevHubRpcError,
  isAbortError,
} from "./errors.js";
import type { NormalizedDevHubClientOptions } from "./models.js";
import { createHttpRequestId, startTimeout, validateResponseEnvelope } from "./jsonrpc.js";
import type { RuntimeConnectionInfo } from "./runtime.js";

export class JsonRpcHttpTransport {
  constructor(
    private readonly options: NormalizedDevHubClientOptions,
    private readonly connection: RuntimeConnectionInfo
  ) {
  }

  async send(method: string, params?: Record<string, unknown> | null): Promise<Record<string, unknown>> {
    if (!method || !method.trim()) {
      throw new Error("method cannot be empty.");
    }

    const requestId = createHttpRequestId();
    const payload: Record<string, unknown> = {
      jsonrpc: "2.0",
      id: requestId,
      method
    };
    if (params !== undefined) {
      payload.params = params;
    }

    const controller = new AbortController();
    const timeoutId = startTimeout(controller, this.options.requestTimeoutMs);

    try {
      const response = await sendHttpRequest(
        this.connection.rpcEndpoint,
        this.connection.token,
        this.options,
        payload,
        controller,
        requestId,
      );

      const body = await readResponseBody(response, controller, requestId);
      if (!response.ok) {
        throw new DevHubConnectionError({
          kind: "http_status",
          message: `HTTP request failed with status ${response.status} ${response.statusText}.`,
          requestId,
          status: response.status,
          statusText: response.statusText,
          responseBody: body
        });
      }

      let root: unknown;
      try {
        root = JSON.parse(body) as unknown;
      } catch (error) {
        throw new DevHubConnectionError({
          kind: "invalid_response",
          message: "Failed to parse JSON-RPC response.",
          cause: error,
          requestId
        });
      }

      try {
        return validateResponseEnvelope(root, requestId);
      } catch (error) {
        if (error instanceof DevHubRpcError) {
          throw error;
        }

        throw new DevHubConnectionError({
          kind: "invalid_response",
          message: error instanceof Error ? error.message : "Invalid JSON-RPC response.",
          cause: error,
          requestId
        });
      }
    } finally {
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    }
  }

  async dispose(): Promise<void> {
  }
}

async function sendHttpRequest(
  endpoint: string,
  token: string,
  options: NormalizedDevHubClientOptions,
  payload: Record<string, unknown>,
  controller: AbortController,
  requestId: string,
): Promise<Response> {
  try {
    return await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
        "X-DevHub-Protocol": String(options.protocolVersion),
        "X-DevHub-ClientId": options.clientId,
        "X-DevHub-ClientSessionId": options.clientSessionId
      },
      body: JSON.stringify(payload),
      signal: controller.signal
    });
  } catch (error) {
    throw mapTransportFailure(error, controller, requestId);
  }
}

async function readResponseBody(
  response: Response,
  controller: AbortController,
  requestId: string,
): Promise<string> {
  try {
    return await response.text();
  } catch (error) {
    throw mapTransportFailure(error, controller, requestId);
  }
}

function mapTransportFailure(
  error: unknown,
  controller: AbortController,
  requestId: string,
): DevHubConnectionError {
  if (controller.signal.aborted || isAbortError(error)) {
    return new DevHubConnectionError({
      kind: "timeout",
      message: "HTTP request timed out.",
      cause: error,
      requestId
    });
  }

  return new DevHubConnectionError({
    kind: "transport",
    message: "HTTP transport failed.",
    cause: error,
    requestId
  });
}
